pilota/src/thrift/binary_unsafe.rs

```diff
@@ -0,0 +1,1186 @@
+use std::{convert::TryInto, ptr, slice, str};
+
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+use faststr::FastStr;
+
+use linkedbytes::LinkedBytes;
+use tokio::io::{AsyncRead, AsyncReadExt};
+
+use super::{
+    error::ProtocolErrorKind, new_protocol_error, DecodeError, DecodeErrorKind, EncodeError,
+    ProtocolError, TAsyncInputProtocol, TFieldIdentifier, TInputProtocol, TLengthProtocol,
+    TListIdentifier, TMapIdentifier, TMessageIdentifier, TMessageType, TOutputProtocol,
+    TSetIdentifier, TStructIdentifier, TType, ZERO_COPY_THRESHOLD,
+};
+
+static VERSION_1: u32 = 0x80010000;
+static VERSION_MASK: u32 = 0xffff0000;
+
+pub struct TBinaryProtocol<T> {
+    pub(crate) trans: T,
+    pub(crate) buf: &'static mut [u8],
+    pub(crate) index: usize,
+
+    zero_copy: bool,
+    zero_copy_len: usize,
+}
+
+impl<T> TBinaryProtocol<T> {
+    /// `zero_copy` only takes effect when `T` is [`BytesMut`] for input and
+    /// [`LinkedBytes`] for output.
+    ///
+    /// The 'buf' must point to the same area of trans, this is a self-referencial struct.
+    #[inline]
+    pub unsafe fn new(trans: T, buf: &'static mut [u8], zero_copy: bool) -> Self {
```

Comment:
Should it better that require `T` pinned？buf would be invalid after T moved. Maybe unsafe function is not necessary.

Reply:
Constraint it with type should be better, if T is pinned, it does not needs to be unsafe

Reply:
The problem here is that, the user need to guarantee that the buf is pointed at trans and trans has enough length here. This cannot be constraint by Pin. This can only be guaranteed by the user.

Reply:
Specify `T: &mut BytesMut` and then slice could be built in `TBinaryProtocol::new` rather than pass it from outside, then we could always get a slice and its length literally equals to `T::len` and it points to buf

Reply:
The `T` type here may also be a `&mut LinkedBytes`, so I can't specify it like that here.

Reply:
`new` method can be implement independently for `TBinaryProtocol<Pin<&mut BytesMut>>` and `TBinaryProtocol<Pin<&mut LinkedBytes>>`